### PR TITLE
Upgrade Multi-Size Page Handling for Enhanced Large Page Support in MSFT OpenJDK on Windows

### PR DESCRIPTION
**Pull request by nitinsingh130519**
This pull request introduces enhancements to the handling of large page sizes in the Microsoft OpenJDK for Windows systems, aiming to align its capabilities with those observed on Linux platforms. Investigation through SPECJBB benchmarks across various platforms revealed a 16-year-old limitation in handling large pages over 4MB for IA32/AMD64 architectures, with no such constraints for Windows on ARM64.

The goal of this change is to overcome the 4MB large page size limitation, thereby enhancing Windows' large page support to match Linux's more flexible handling capabilities. This decision was influenced by issues observed with code cache and large page utilization, alongside insights from Linux's implementation strategies. The implementation supports multiple large page sizes on recent versions of Windows (Windows Client version >=11 and Windows Server version >=22), specifically excluding the IA32 architecture.

Key changes and bug fixes include enabling Linux support for multiple huge page sizes with -XX:LargePageSizeInBytes, adapting similar logic for Windows support, and utilizing logic from JDK-8271195 to use the largest available large page size smaller than LargePageSizeInBytes when available. This approach is further supported by addressing several JDK bug reports and enhancement requests related to large page handling, ensuring a comprehensive update to large page management.

This update removes the 4MB limit on AMD64 for Windows 11+ and Server 22+, populating the shared array to enable fallback options on all architectures except IA32. The implementation introduces an experimental flag, defaulting to FALSE, to facilitate testing and gradual adoption of these changes. The flag allows users to opt-in to the new large page handling logic, with a warning mechanism implemented for cases where the requested large page size is not a multiple of the OS minimum page size.

### CODE REVIEW

File: src/hotspot/os/windows/os_windows.cpp [hunk 1]
```diff
@@ -3241,15 +3242,33 @@ static size_t large_page_init_decide_size() {
     return 0;
   }
 
-#if defined(IA32) || defined(AMD64)
-  if (size > 4*M || LargePageSizeInBytes > 4*M) {
-    WARN("JVM cannot use large pages bigger than 4mb.");
-    return 0;
-  }
+  // Check if OS version is 11 or greater
+  if (schedules_all_processor_groups) {
```
Comment (macarte) [inline src/hotspot/os/windows/os_windows.cpp:3246]: either have saint change the name to something more generic or call instead "win32::is_windows_11_or_greater() || win32::is_windows_server_2022_or_greater()"

File: src/hotspot/os/windows/os_windows.cpp [hunk 2]
```diff
@@ -3265,8 +3284,23 @@ void os::large_page_init() {
   _large_page_size = large_page_init_decide_size();
   const size_t default_page_size = os::vm_page_size();
   if (_large_page_size > default_page_size) {
-    _page_sizes.add(_large_page_size);
+ #if !defined(IA32)
+      // Check if the OS version is 11 or higher
+      if (schedules_all_processor_groups && EnableAllLargePageSizes) {
```
Comment (macarte) [inline src/hotspot/os/windows/os_windows.cpp:3289]: This is OS independent right? If so only check the new JVM flag

File: src/hotspot/os/windows/os_windows.cpp [hunk 3]
```diff
@@ -3265,8 +3284,23 @@ void os::large_page_init() {
...
+
+          // Additional logic needed for ARM architecture
+          size_t min_size = GetLargePageMinimum();
+
+          // Populate _page_sizes with large page sizes less than or equal to _large_page_size.
```
Comment (macarte) [inline src/hotspot/os/windows/os_windows.cpp:3294]: .... each page is double the previous

File: test/hotspot/gtest/runtime/test_os_windows.cpp [hunk 4]
```diff
@@ -703,6 +703,54 @@ TEST_VM(os_windows, reserve_memory_special) {
   TestReserveMemorySpecial_test();
 }
 
+TEST_VM(os_windows, large_page_init_decide_size) {
```
Comment (macarte) [inline test/hotspot/gtest/runtime/test_os_windows.cpp:706]: should we also have a test that there are multiple page sizes in the list and that they are all multiples of the smallest and do not exceed the maxlargepagesize?

File: src/hotspot/os/windows/os_windows.cpp [hunk 5]
```diff
@@ -3241,15 +3242,33 @@ static size_t large_page_init_decide_size() {
...
 
   if (LargePageSizeInBytes > 0 && LargePageSizeInBytes % size == 0) {
-    size = LargePageSizeInBytes;
+      size = LargePageSizeInBytes;
+  }
+  else {
+      WARN1("JVM cannot use large pages that are not a multiple of minimum large page size (%d), defaulting to minimum page size.", size);
   }
 
 #undef WARN
```
Comment (macarte) [inline src/hotspot/os/windows/os_windows.cpp:3274]: we also need to #undef WARN1

**Comment by nitinsingh130519** [comment]
Ignore this PR, mistakenly added so many commits

**Pull request closed without merging**